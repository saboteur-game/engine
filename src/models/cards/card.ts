import { generateId, Pojo } from "../../utils";
import CardParameters from "./card-parameters";

export enum Status {
  unused = "unused",
  played = "played",
  discarded = "discarded",
}

class Card {
  id: string;
  status: Status;

  constructor() {
    this.id = generateId();
    this.status = Status.unused;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  play(parameters: CardParameters): Card {
    return this;
  }

  setDiscarded(): void {
    // TODO: Check if the state was unused and throw if it was anything else
    this.status = Status.discarded;
  }

  setPlayed(): void {
    // TODO: Check if the state was unused or played and throw if it was anything else
    this.status = Status.played;
  }

  toJS(): Pojo {
    return {
      id: this.id,
      status: this.status,
    };
  }
}

export default Card;
