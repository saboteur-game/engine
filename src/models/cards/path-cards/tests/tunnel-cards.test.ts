import { Sides } from "../../card";
import { PassageCard, DeadendCard } from "..";

describe("Tunnel Cards", () => {
  describe("PassageCard", () => {
    const passageConfigurations = [
      [[Sides.right, Sides.left]],
      [[Sides.top, Sides.bottom]],
      [[Sides.top, Sides.left]],
      [[Sides.top, Sides.right]],
      [[Sides.top, Sides.right, Sides.left]],
      [[Sides.top, Sides.right, Sides.bottom]],
      [[Sides.top, Sides.right, Sides.bottom, Sides.left]],
    ];

    describe.each([
      ["right side up", false],
      ["upside down", true],
    ])("%s", (message, direction) => {
      describe.each(passageConfigurations)(
        "with %p configuration",
        (connections) => {
          it("can be visualized", () => {
            expect(
              new PassageCard(connections, direction).visualize()
            ).toMatchSnapshot();
          });
        }
      );
    });
  });

  describe("DeadendCard", () => {
    const deadendConfigurations = [
      [[Sides.top]],
      [[Sides.right]],
      [[Sides.top, Sides.right]],
      [[Sides.top, Sides.bottom]],
      [[Sides.top, Sides.left]],
      [[Sides.right, Sides.left]],
      [[Sides.top, Sides.right, Sides.bottom]],
      [[Sides.top, Sides.right, Sides.left]],
      [[Sides.top, Sides.right, Sides.bottom, Sides.left]],
    ];

    describe.each([
      ["right side up", false],
      ["upside down", true],
    ])("%s", (message, direction) => {
      describe.each(deadendConfigurations)(
        "with %p configuration",
        (connections) => {
          it("can be visualized", () => {
            expect(
              new DeadendCard(connections, direction).visualize()
            ).toMatchSnapshot();
          });
        }
      );
    });
  });
});
