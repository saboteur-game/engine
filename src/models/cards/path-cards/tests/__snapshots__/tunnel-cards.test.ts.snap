// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tunnel Cards DeadendCard right side up with [1, 2, 3, 4] configuration can be visualized 1`] = `
"██  ██
  ██  
██  ██"
`;

exports[`Tunnel Cards DeadendCard right side up with [1, 2, 3] configuration can be visualized 1`] = `
"██  ██
████  
██  ██"
`;

exports[`Tunnel Cards DeadendCard right side up with [1, 2, 4] configuration can be visualized 1`] = `
"██  ██
  ██  
██████"
`;

exports[`Tunnel Cards DeadendCard right side up with [1, 2] configuration can be visualized 1`] = `
"██  ██
████  
██████"
`;

exports[`Tunnel Cards DeadendCard right side up with [1, 3] configuration can be visualized 1`] = `
"██  ██
██████
██  ██"
`;

exports[`Tunnel Cards DeadendCard right side up with [1, 4] configuration can be visualized 1`] = `
"██  ██
  ████
██████"
`;

exports[`Tunnel Cards DeadendCard right side up with [1] configuration can be visualized 1`] = `
"██  ██
██████
██████"
`;

exports[`Tunnel Cards DeadendCard right side up with [2, 4] configuration can be visualized 1`] = `
"██████
  ██  
██████"
`;

exports[`Tunnel Cards DeadendCard right side up with [2] configuration can be visualized 1`] = `
"██████
████  
██████"
`;

exports[`Tunnel Cards DeadendCard upside down with [1, 2, 3, 4] configuration can be visualized 1`] = `
"██  ██
  ██  
██  ██"
`;

exports[`Tunnel Cards DeadendCard upside down with [1, 2, 3] configuration can be visualized 1`] = `
"██  ██
  ████
██  ██"
`;

exports[`Tunnel Cards DeadendCard upside down with [1, 2, 4] configuration can be visualized 1`] = `
"██████
  ██  
██  ██"
`;

exports[`Tunnel Cards DeadendCard upside down with [1, 2] configuration can be visualized 1`] = `
"██████
  ████
██  ██"
`;

exports[`Tunnel Cards DeadendCard upside down with [1, 3] configuration can be visualized 1`] = `
"██  ██
██████
██  ██"
`;

exports[`Tunnel Cards DeadendCard upside down with [1, 4] configuration can be visualized 1`] = `
"██████
████  
██  ██"
`;

exports[`Tunnel Cards DeadendCard upside down with [1] configuration can be visualized 1`] = `
"██████
██████
██  ██"
`;

exports[`Tunnel Cards DeadendCard upside down with [2, 4] configuration can be visualized 1`] = `
"██████
  ██  
██████"
`;

exports[`Tunnel Cards DeadendCard upside down with [2] configuration can be visualized 1`] = `
"██████
  ████
██████"
`;

exports[`Tunnel Cards PassageCard right side up with [1, 2, 3, 4] configuration can be visualized 1`] = `
"██  ██
      
██  ██"
`;

exports[`Tunnel Cards PassageCard right side up with [1, 2, 3] configuration can be visualized 1`] = `
"██  ██
██    
██  ██"
`;

exports[`Tunnel Cards PassageCard right side up with [1, 2, 4] configuration can be visualized 1`] = `
"██  ██
      
██████"
`;

exports[`Tunnel Cards PassageCard right side up with [1, 2] configuration can be visualized 1`] = `
"██  ██
██    
██████"
`;

exports[`Tunnel Cards PassageCard right side up with [1, 3] configuration can be visualized 1`] = `
"██  ██
██  ██
██  ██"
`;

exports[`Tunnel Cards PassageCard right side up with [1, 4] configuration can be visualized 1`] = `
"██  ██
    ██
██████"
`;

exports[`Tunnel Cards PassageCard right side up with [2, 4] configuration can be visualized 1`] = `
"██████
      
██████"
`;

exports[`Tunnel Cards PassageCard upside down with [1, 2, 3, 4] configuration can be visualized 1`] = `
"██  ██
      
██  ██"
`;

exports[`Tunnel Cards PassageCard upside down with [1, 2, 3] configuration can be visualized 1`] = `
"██  ██
    ██
██  ██"
`;

exports[`Tunnel Cards PassageCard upside down with [1, 2, 4] configuration can be visualized 1`] = `
"██████
      
██  ██"
`;

exports[`Tunnel Cards PassageCard upside down with [1, 2] configuration can be visualized 1`] = `
"██████
    ██
██  ██"
`;

exports[`Tunnel Cards PassageCard upside down with [1, 3] configuration can be visualized 1`] = `
"██  ██
██  ██
██  ██"
`;

exports[`Tunnel Cards PassageCard upside down with [1, 4] configuration can be visualized 1`] = `
"██████
██    
██  ██"
`;

exports[`Tunnel Cards PassageCard upside down with [2, 4] configuration can be visualized 1`] = `
"██████
      
██████"
`;
